imt/tree_writer.go

```diff
@@ -75,27 +71,24 @@ func (t *TreeWriter) Insert(key, value *big.Int) (*MutateProof, error) {
 	}
 	size++
 
-	err = t.setKeyIndex(key, size)
-	if err != nil {
-		return nil, err
-	}
 	err = t.setSize(size)
 	if err != nil {
 		return nil, err
 	}
 
 	newNode := &Node{
 		Key:     key,
+		Index:   size,
 		Value:   value,
 		NextKey: lowNode.NextKey,
 	}
-	_, err = t.setNode(size, newNode)
+	_, err = t.setNode(newNode)
 	if err != nil {
 		return nil, err
 	}
 
 	lowNode.NextKey = key
```

Comment:
We might have a reference issue here. 

In the previous implementation a new node struct is instantiated each time the `t.ProveIndex` method is called and the generated proof is pointing to this newly allocated struct, no problem.

In the new implementation, `t.ProveNode` is accepting an existing `Node` reference. Once a `Proof` is generated and pointing to its associated `Node` we need to make sure that this `Node` is not updated after (else the proof would fail to validate in the `Valid` method). This is happening here I think as `oldProof` still points to `lowNode` but we are updating it here.

Here are the few options I'm thinking of:
1. Deep copy the node in `t.ProveNode`
2. Or directly store the node fields in the `Proof`

Reply:
Great catch. This would not have affected anything because the `oldProof`'s node is not used anywhere (only the `oldProof.siblings`). But made a clone of the node anyway 👍

Reply:
Got it. I would personally have done the cloning in the proving method (`ProveNode`) so that the proof is responsible of the cloned version and this issue does not come back later if you add more logic. But I don't see any problem with doing it the way you did.

Reply:
On second thought, i removed the cloning altogther, and replaced the call to `ProveNode` with one that only provides the siblings for the proof, which seems cleaner to me